File: cats.py
import json
import requests

# Fungsi untuk membaca file query.txt yang berisi header authorization untuk banyak akun
def read_authorizations(file_path):
    with open(file_path, 'r') as f:
        lines = f.readlines()
    return [line.strip() for line in lines]

# Fungsi untuk mendapatkan informasi user dan menampilkan data username dan totalRewards
def process_user_info(authorization_header, account_number):
    headers = {
        "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, seperti Gecko) Chrome/120.0.0.0 Mobile Safari/537.36",
        "Content-Type": "application/json",
        "Authorization": f"tma {authorization_header}",
        "Origin": "https://cats-frontend.tgapps.store"
    }

    response = requests.get("https://api.catshouse.club/user", headers=headers)

    if response.status_code == 200:
        data = response.json()
        username = data.get('username')
        total_rewards = data.get('totalRewards')
        print(f"AKUN KE#{account_number} - {username} {total_rewards} CATS")
    else:
        print(f"Error fetching user info for account {account_number}: {response.status_code}")

# Fungsi untuk get daftat / list tasks yabg terdia dan klaim jika belum completed
def process_tasks(authorization_header):
    headers = {
        "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, seperti Gecko) Chrome/120.0.0.0 Mobile Safari/537.36",
        "Content-Type": "application/json",
        "Authorization": f"tma {authorization_header}",
        "Origin": "https://cats-frontend-production.pages.dev"
    }

    response = requests.get("https://api.catshouse.club/tasks/user", headers=headers)

    if response.status_code == 200:
        tasks = response.json().get("tasks", [])
        for task in tasks:
            task_id = task.get('id')
            title = task.get('title')
            reward_points = task.get('rewardPoints')
            completed = task.get('completed')

            # Cek status completed dan klaim jika false
            status = "ALREADY CLAIM!!" if completed else "NOT CLAIMED!!"
            print(f"#ID {task_id} - {title} - GET {reward_points}CATS - {status}")

            if not completed:
                claim_task(authorization_header, task_id)
    else:
        print(f"[!] QUERY LU MOKAD BOSSKU!!")

# Fungsi untuk klaim task yang tersedia
def claim_task(authorization_header, task_id):
    headers = {
        "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, seperti Gecko) Chrome/120.0.0.0 Mobile Safari/537.36",
        "Content-Type": "application/json",
        "Authorization": f"tma {authorization_header}",
        "Origin": "https://cats-frontend-production.pages.dev"
    }

    url = f"https://api.catshouse.club/tasks/{task_id}/complete"
    response = requests.post(url, headers=headers, json={})

    if response.status_code == 200:
        result = response.json()
        if 'success' in result and result['success']:
            print("BERHASIL CLAIM!!")
        else:
            print("GAGAL CLAIM / SUDAH DICLAIM SEBELUMNYA!!")
    else:
        # Handling manual claim errors / harus manual claim
        result = response.json()
        if 'name' in result and result['name'] == "Error":
            print("SKIP TASK - MANUAL CLAIM!!")
        else:
            print(f"[!] QUERY LU MOKAD BOSSKU!! ")


# Main function
def main():
    # Baca daftar query yang tersedia dari file query.txt
    authorization_headers = read_authorizations('query.txt')
    
    # Proses setiap akun yang tersedia pada file query.txt
    for account_number, authorization_header in enumerate(authorization_headers, start=1):
        
        # Proses get informasi user with query
        process_user_info(authorization_header, account_number)

        # Proses get task dan claim tasks
        process_tasks(authorization_header)

if __name__ == "__main__":
    main()

File: cats.js
import fetch from 'node-fetch';
import fs from 'fs';

function getToken(filename) {
    const data = fs.readFileSync(filename, 'utf8');
    return data.split('\n').map(line => line.trim()).filter(line => line.length > 0);
}
    
function number(number, decimals = 0, decPoint = ',', thousandsSep = '.') {
    const n = parseFloat(number).toFixed(decimals);
    const parts = n.split('.');
    parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandsSep);
    return parts.join(decPoint);
}

async function getCURL(url, method = 'GET', headers = {}, body = null, returnJson = true) {
    const options = {
        method,
        headers,
    };

    if (body && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
        options.body = JSON.stringify(body);
    }

    const response = await fetch(url, options);
    const data = returnJson ? await response.json() : await response.text();
    
    return data;
}

(async () => {
    const dataList = getToken('dataAkun.txt');
    
    console.log(`-------------------------------`);
    console.log(` |            MENU            | `);
    console.log(` [  CATS.BOT AUTO CLEAR TASK  ] `);
    console.log(`-------------------------------`);
    console.log();

    const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
    console.log('[.] MENJALANKAN AUTO CLEAR TASK, DELAY 24 JAM SETELAH CEK ' + dataList.length + ' AKUN...\n');
    while (true) {
        for (let i = 0; i < dataList.length; i += 1) {
            const batch = dataList.slice(i, i + 1);
            const batchPromises = batch.map(async (token, batchIndex) => {
            const no = i + batchIndex + 1;
            // Parsing query string menggunakan URLSearchParams
            const params = new URLSearchParams(token);
            const user = JSON.parse(decodeURIComponent(params.get('user')));
            let logMessage = `====================================================
[[#${no}] MENGAMBIL DATA AKUN: ${user.username} ]\n`;
            const headers = {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0',
                'Accept-Encoding': 'gzip, deflate, br, zstd',
                 'authorization': `tma ${token}`,
                'origin': 'https://api.catshouse.club'
            };
                const infoAkun = await getCURL('https://api.catshouse.club/user', 'GET', headers);
                if (infoAkun.name == 'Error') {
                    logMessage += `[x] TOKEN QUERY_ID MOKAD!!\n`;

                } else {
                    logMessage += `
[ #.NAME ] : ${infoAkun.firstName ? infoAkun.firstName : ""} ${infoAkun.lastName ? infoAkun.lastName : ""} ( ${infoAkun.id} )
[ #.BALLANCE ] : ${number(infoAkun.totalRewards)} CATS
[ #.TG AGE ] : ${number(infoAkun.telegramAge)}\n
[*] INFORMASI CLAIM TASK:\n`;

                const infoClaim = await getCURL('https://api.catshouse.club/tasks/user', 'GET', headers);
                if (infoClaim) {
                    if (infoClaim.tasks && Array.isArray(infoClaim.tasks)) {
                        for (const task of infoClaim.tasks) {
                            if (task.completed === false) {
                                // Auto claim untuk setiap tugas
                                try {
                                    const claimTask = await getCURL(`https://api.catshouse.club/tasks/${task.id}/complete`, 'POST', headers, {});
                                    
                                    if (claimTask.success === true) {
                                        const infoAkun = await getCURL('https://api.catshouse.club/user', 'GET', headers);
                                        logMessage += `[#] CLAIM ${task.title.toUpperCase()} GET ${number(task.rewardPoints)} CATS => BERHASIL!! - `;
                                        logMessage += `${number(infoAkun.totalRewards)} CATS\n`;
                                    }else if (claimTask.success === false) {
                                        logMessage += `[#] CLAIM ${task.title.toUpperCase()} GET ${number(task.rewardPoints)} CATS => SKIPP!! NEED JOIN OR MANUAL!!!!\n`;
                                    }
                                } catch (error) {
                                    logMessage += `[!] ERROR CLAIMING TASK ID ${task.id}: ${error.message}\n`;
                                }
                            }
                            
                        }
                
                logMessage += `\n[#] SEMUA TASK BERHASIL DICLAIM!! MENUNGGU TASK BARU!!\n`;

                    }
                } else if (infoClaim && infoClaim.name === 'Error') {
                    logMessage += `[x] CLAIM TOKEN MOKAD!!!!\n`;
                } else {
                    logMessage += `[x] ERROR!! YGTKTS!!\n`;
                }

                }
                console.log(logMessage);
            });
            await Promise.all(batchPromises);
        }
        console.log(`[${getCurrentTime()}] SEMUA AKUN BERHASIL DIPROSESS, DELAY 24 JAM...`);
        await delay(87000);
        console.clear();
        console.log(`[${getCurrentTime()}] MEMULAI AUTO CLAIM ${dataList.length} AKUN...\n`);
    }
})();

function getCurrentTime() {
    const now = new Date();
    const options = {
        timeZone: 'Asia/Jakarta',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hour12: false
    };

    const timeFormatter = new Intl.DateTimeFormat('en-GB', options);
    const timeParts = timeFormatter.formatToParts(now);

    const hours = timeParts.find(part => part.type === 'hour').value;
    const minutes = timeParts.find(part => part.type === 'minute').value;
    const seconds = timeParts.find(part => part.type === 'second').value;

    return `${hours}:${minutes}:${seconds}`;
}
